### PR TITLE
Fix force unwrap in BaseCollectionViewController

### PR DESCRIPTION
**Pull request by alaija**
Force unwrap leads to crash in cases when ViewController addresses to collectionview before it was loaded. e.g. You can try to rotate device when chat is initialized inside unopened tab inside UITabBarController.

### CODE REVIEW

**Comment by alaija** [comment]
Actually, we made this(isViewLoaded) workaround in our child viewcontroller, but it looks like workaround and not a solution, cause we speak about “Base” class and someone who will implement his child viewcontroller can be faced to the problem again.

Anyway layouting is just one case, there are too many of them, and I really surprised that no one met any before me, so in my opinion optional views is the only solution.

File: Chatto/Source/ChatController/BaseChatViewController+Scrolling.swift [hunk 1]
```diff
@@ -36,79 +36,97 @@ extension CGFloat {
 extension BaseChatViewController {
 
     public func isScrolledAtBottom() -> Bool {
-        guard self.collectionView.numberOfSections > 0 && self.collectionView.numberOfItems(inSection: 0) > 0 else { return true }
-        let sectionIndex = self.collectionView.numberOfSections - 1
-        let itemIndex = self.collectionView.numberOfItems(inSection: sectionIndex) - 1
+        guard
```
Comment (AntonPalich) [inline Chatto/Source/ChatController/BaseChatViewController+Scrolling.swift]: Could I ask you to use several guard statements here and below ? It would be more readable:

```
guard let collectionView = self.collectionView else { return true }
guard let ... else { return true }
```

File: Chatto/Source/ChatController/BaseChatViewController+Changes.swift [hunk 2]
```diff
@@ -71,13 +71,14 @@ extension BaseChatViewController {
             return 0
         }
 
-        let contentHeight = self.collectionView.contentSize.height
```
Comment (AntonPalich) [inline Chatto/Source/ChatController/BaseChatViewController+Changes.swift]: The most of functions that access `collectionView` don't make sense to call if it wasn't created yet. Could it be better if you use `guard` statement at the beginning to get unwrapped `collectionView`?

File: Chatto/Source/ChatController/BaseChatViewController+Scrolling.swift [hunk 3]
```diff
@@ -36,79 +36,97 @@ extension CGFloat {
...
+        return (self.visibleRect().minY / collectionView.contentSize.height) < self.constants.autoloadingFractionalThreshold
     }
 
     public func isIndexPathVisible(_ indexPath: IndexPath, atEdge edge: CellVerticalEdge) -> Bool {
-        if let attributes = self.collectionView.collectionViewLayout.layoutAttributesForItem(at: indexPath) {
+        guard
+            let collectionView = self.collectionView,
+            let attributes = collectionView.collectionViewLayout.layoutAttributesForItem(at: indexPath)
+        else { return false }
             let visibleRect = self.visibleRect()
```
Comment (AntonPalich) [inline Chatto/Source/ChatController/BaseChatViewController+Scrolling.swift]: Those lines have wrong indentation after `if` statement was removed.

Comment (alaija) [reply Chatto/Source/ChatController/BaseChatViewController+Scrolling.swift]: Sorry, It happens when swiftformat fixes it for you on the main project.

File: Chatto/Source/ChatController/BaseChatViewController+Scrolling.swift (hunk 1 above)
Comment (alaija) [reply Chatto/Source/ChatController/BaseChatViewController+Scrolling.swift]: I think it easier to read when you have only return statement, but OK

File: Chatto/Source/ChatController/BaseChatViewController+Changes.swift (hunk 2 above)
Comment (alaija) [reply Chatto/Source/ChatController/BaseChatViewController+Changes.swift]: Done

File: Chatto/Source/ChatController/BaseChatViewController+Changes.swift [hunk 4]
```diff
@@ -128,7 +130,10 @@ extension BaseChatViewController {
                              changes: CollectionChanges,
                              updateType: UpdateType,
                              completion: @escaping () -> Void) {
-
+        guard let collectionView = self.collectionView else {
```
Comment (alaija) [inline Chatto/Source/ChatController/BaseChatViewController+Changes.swift]: Please check this method carefully, cause it too complicated to get what blocks should I call on exit.

P.S. It is a good candidate for refactoring

Comment (AntonPalich) [reply Chatto/Source/ChatController/BaseChatViewController+Changes.swift]: There are many places that requires refactoring :) Someday we will take care about them

File: Chatto/Source/ChatController/BaseChatViewController.swift [hunk 5]
```diff
@@ -230,7 +231,6 @@ open class BaseChatViewController: UIViewController, UICollectionViewDataSource,
     }
 
     open func handleKeyboardPositionChange(bottomMargin: CGFloat, keyboardStatus: KeyboardStatus) {
-        guard self.inputContainerBottomConstraint.constant != bottomMargin else { return }
```
Comment (AntonPalich) [inline Chatto/Source/ChatController/BaseChatViewController.swift]: Please, revert this change

File: Chatto/Source/ChatController/BaseChatViewController.swift [hunk 6]
```diff
@@ -391,9 +393,11 @@ open class BaseChatViewController: UIViewController, UICollectionViewDataSource,
 extension BaseChatViewController { // Rotation
 
     open override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
+        guard isViewLoaded else { return }
+        guard let collectionView = self.collectionView else { return }
         super.viewWillTransition(to: size, with: coordinator)
```
Comment (AntonPalich) [inline Chatto/Source/ChatController/BaseChatViewController.swift]: I believe, `super` implementation should be called anyway.
